File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/tools/LiteFileTools.java
package com.zzf.rikki.idea.agent.tools;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.vfs.LocalFileSystem;

import java.io.BufferedReader;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

public class LiteFileTools {
    private static final int DEFAULT_READ_LIMIT = 2_000;
    private static final int MAX_LINE_LEN = 2_000;
    private static final int MAX_BYTES = 50 * 1024;
    private static final int MAX_OUTPUT = 8_000;
    private static final List<String> BINARY_EXTS = List.of(
            "zip", "tar", "gz", "exe", "dll", "so", "class", "jar", "war", "7z", "bin", "dat",
            "png", "jpg", "jpeg", "gif", "bmp", "ico", "pdf", "mp3", "mp4", "avi"
    );

    public LiteFileTools(ObjectMapper mapper) {
    }

    public String read(JsonNode args, String workspaceRoot) throws Exception {
        String pathStr = args.path("filePath").asText("");
        int offset = Math.max(args.path("offset").asInt(0), 0);
        int limit = Math.max(args.path("limit").asInt(DEFAULT_READ_LIMIT), 1);
        File file = resolve(workspaceRoot, pathStr);
        if (!file.exists()) {
            throw new RuntimeException("File not found: " + pathStr);
        }
        if (isBinary(file)) {
            throw new RuntimeException("Cannot read binary file: " + pathStr);
        }
        List<String> allLines = Files.readAllLines(file.toPath(), StandardCharsets.UTF_8);
        StringBuilder builder = new StringBuilder("<file>\n");
        int bytes = 0;
        boolean truncatedByBytes = false;
        int end = Math.min(allLines.size(), offset + limit);
        for (int i = offset; i < end; i++) {
            String line = allLines.get(i);
            if (line.length() > MAX_LINE_LEN) {
                line = line.substring(0, MAX_LINE_LEN) + "...";
            }
            int size = line.getBytes(StandardCharsets.UTF_8).length + 1;
            if (bytes + size > MAX_BYTES) {
                truncatedByBytes = true;
                break;
            }
            builder.append(String.format("%05d| %s%n", i + 1, line));
            bytes += size;
        }
        int totalLines = allLines.size();
        if (truncatedByBytes) {
            builder.append("\n\n(Truncated at ").append(MAX_BYTES).append(" bytes. Use 'offset' to continue.)");
        } else if (totalLines > offset + limit) {
            builder.append("\n\n(More lines exist. Use 'offset' to read beyond line ").append(offset + limit).append(".)");
        } else {
            builder.append("\n\n(End of file - ").append(totalLines).append(" lines total)");
        }
        builder.append("\n</file>");
        return builder.toString();
    }

    public String write(JsonNode args, String workspaceRoot) throws Exception {
        String pathStr = args.path("filePath").asText("");
        String content = args.path("content").asText("");
        File file = resolve(workspaceRoot, pathStr);
        if (file.getParentFile() != null) {
            file.getParentFile().mkdirs();
        }
        Files.writeString(file.toPath(), content, StandardCharsets.UTF_8);
        refreshVfsSync(file.getAbsolutePath());
        return "Written: " + pathStr + " (" + content.lines().count() + " lines)";
    }

    public String edit(JsonNode args, String workspaceRoot) throws Exception {
        String pathStr = args.path("filePath").asText("");
        String oldString = args.path("oldString").asText("");
        String newString = args.path("newString").asText("");
        boolean replaceAll = args.path("replaceAll").asBoolean(false);
        File file = resolve(workspaceRoot, pathStr);
        if (!file.exists()) {
            if (!oldString.isEmpty()) {
                throw new RuntimeException("File not found: " + pathStr + ". To create, leave oldString empty.");
            }
            if (file.getParentFile() != null) {
                file.getParentFile().mkdirs();
            }
            Files.writeString(file.toPath(), newString, StandardCharsets.UTF_8);
            refreshVfsSync(file.getAbsolutePath());
            return "Created: " + pathStr;
        }
        if (oldString.equals(newString)) {
            throw new RuntimeException("oldString and newString must differ");
        }
        String original = Files.readString(file.toPath(), StandardCharsets.UTF_8);
        List<String> matches = findMatches(original, oldString);
        if (matches.isEmpty()) {
            throw new RuntimeException("oldString not found in file (exact or trimmed match)");
        }
        if (!replaceAll && matches.size() > 1) {
            throw new RuntimeException("oldString found " + matches.size() + " times - provide more context to uniquely identify the match");
        }
        String updated = original;
        if (replaceAll) {
            for (String match : matches) {
                updated = updated.replace(match, newString);
            }
        } else {
            updated = updated.replace(matches.get(0), newString);
        }
        Files.writeString(file.toPath(), updated, StandardCharsets.UTF_8);
        refreshVfsSync(file.getAbsolutePath());
        return "Updated: " + pathStr;
    }

    public String delete(JsonNode args, String workspaceRoot) throws Exception {
        String pathStr = args.path("filePath").asText("");
        File file = resolve(workspaceRoot, pathStr);
        if (!file.exists()) {
            throw new RuntimeException("File not found: " + pathStr);
        }
        if (!file.delete()) {
            throw new RuntimeException("Failed to delete: " + pathStr);
        }
        refreshVfsSync(file.getAbsolutePath());
        return "Deleted: " + pathStr;
    }

    public String glob(JsonNode args, String workspaceRoot) throws Exception {
        String pattern = args.path("pattern").asText("");
        if (pattern.isBlank()) {
            throw new IllegalArgumentException("pattern required");
        }
        File searchRoot = resolve(workspaceRoot, args.path("path").asText(workspaceRoot));
        var matcher = FileSystems.getDefault().getPathMatcher("glob:" + pattern);
        List<PathWithTime> results = new ArrayList<>();
        try (var walk = Files.walk(searchRoot.toPath())) {
            walk.filter(path -> {
                Path rel = searchRoot.toPath().relativize(path);
                return matcher.matches(rel) || matcher.matches(path.getFileName());
            }).forEach(path -> results.add(new PathWithTime(path.toFile().lastModified(), path.toString())));
        }
        results.sort(Comparator.comparingLong(PathWithTime::mtime).reversed());
        List<String> lines = results.stream().limit(100).map(PathWithTime::path).toList();
        return lines.isEmpty() ? "No files matched pattern: " + pattern : String.join("\n", lines);
    }

    public String grep(JsonNode args, String workspaceRoot) throws Exception {
        String patternStr = args.path("pattern").asText("");
        if (patternStr.isBlank()) {
            throw new IllegalArgumentException("pattern required");
        }
        File searchRoot = resolve(workspaceRoot, args.path("path").asText(workspaceRoot));
        String include = args.path("include").asText("");
        var includeMatcher = include.isBlank() ? null : FileSystems.getDefault().getPathMatcher("glob:" + include);
        Pattern compiledRegex;
        try {
            compiledRegex = Pattern.compile(patternStr);
        } catch (PatternSyntaxException ignored) {
            compiledRegex = Pattern.compile(Pattern.quote(patternStr));
        }
        final Pattern regex = compiledRegex;
        List<PathWithTime> results = new ArrayList<>();
        try (var walk = Files.walk(searchRoot.toPath())) {
            walk.filter(Files::isRegularFile).forEach(path -> {
                try {
                    File file = path.toFile();
                    if (isBinary(file)) {
                        return;
                    }
                    if (includeMatcher != null && !includeMatcher.matches(path.getFileName())) {
                        return;
                    }
                    try (BufferedReader reader = Files.newBufferedReader(path, StandardCharsets.UTF_8)) {
                        String line;
                        int lineNum = 0;
                        while ((line = reader.readLine()) != null) {
                            lineNum++;
                            if (regex.matcher(line).find()) {
                                results.add(new PathWithTime(file.lastModified(), path + ":" + lineNum + ": " + line));
                            }
                        }
                    }
                } catch (Exception ignored) {
                }
            });
        }
        results.sort(Comparator.comparingLong(PathWithTime::mtime).reversed());
        List<String> lines = results.stream().limit(100).map(PathWithTime::path).toList();
        String output = lines.isEmpty() ? "No matches for pattern: " + patternStr : String.join("\n", lines);
        return output.length() > MAX_OUTPUT ? output.substring(0, MAX_OUTPUT) : output;
    }

    public String list(JsonNode args, String workspaceRoot) {
        File dir = resolve(workspaceRoot, args.path("path").asText(workspaceRoot));
        if (!dir.isDirectory()) {
            throw new RuntimeException("Not a directory: " + dir.getPath());
        }
        List<String> ignorePatterns = new ArrayList<>();
        if (args.has("ignore") && args.path("ignore").isArray()) {
            args.path("ignore").forEach(node -> ignorePatterns.add(node.asText("")));
        }
        ignorePatterns.addAll(List.of(".git", "node_modules", ".gradle", "build", "out", ".idea", "__pycache__"));
        StringBuilder builder = new StringBuilder();
        walk(dir, "", ignorePatterns, builder);
        return builder.length() == 0 ? "Empty directory" : builder.toString().trim();
    }

    private void walk(File dir, String indent, List<String> ignorePatterns, StringBuilder builder) {
        if (builder.length() > MAX_OUTPUT) {
            return;
        }
        File[] children = dir.listFiles();
        if (children == null) {
            return;
        }
        List<File> sorted = new ArrayList<>(List.of(children));
        sorted.sort(Comparator.<File, Boolean>comparing(file -> !file.isDirectory()).thenComparing(File::getName));
        for (File child : sorted) {
            if (shouldIgnore(child.getName(), ignorePatterns)) {
                continue;
            }
            builder.append(indent).append(child.getName());
            if (child.isDirectory()) {
                builder.append("/");
            }
            builder.append("\n");
            if (child.isDirectory()) {
                walk(child, indent + "  ", ignorePatterns, builder);
            }
        }
    }

    private boolean shouldIgnore(String name, List<String> ignorePatterns) {
        for (String pattern : ignorePatterns) {
            if (name.equals(pattern)) {
                return true;
            }
            if (name.startsWith(".") && ".*".equals(pattern)) {
                return true;
            }
        }
        return false;
    }

    private List<String> findMatches(String content, String find) {
        if (find.isEmpty()) {
            return List.of();
        }
        List<String> results = new ArrayList<>();
        String[] origLines = content.split("\n", -1);
        String[] searchLinesRaw = find.split("\n", -1);
        List<String> searchLines = new ArrayList<>(List.of(searchLinesRaw));
        if (!searchLines.isEmpty() && searchLines.get(searchLines.size() - 1).isEmpty()) {
            searchLines.remove(searchLines.size() - 1);
        }
        if (searchLines.isEmpty() || origLines.length < searchLines.size()) {
            return results;
        }
        for (int i = 0; i <= origLines.length - searchLines.size(); i++) {
            boolean matches = true;
            for (int j = 0; j < searchLines.size(); j++) {
                if (!origLines[i + j].trim().equals(searchLines.get(j).trim())) {
                    matches = false;
                    break;
                }
            }
            if (matches) {
                int start = 0;
                for (int k = 0; k < i; k++) {
                    start += origLines[k].length() + 1;
                }
                int end = start;
                for (int k = 0; k < searchLines.size(); k++) {
                    end += origLines[i + k].length();
                    if (k < searchLines.size() - 1) {
                        end += 1;
                    }
                }
                results.add(content.substring(start, end));
            }
        }
        return results;
    }

    private void refreshVfsSync(String absolutePath) {
        try {
            String normalized = absolutePath.replace('\\', '/');
            ApplicationManager.getApplication().invokeAndWait(() -> {
                var file = LocalFileSystem.getInstance().refreshAndFindFileByPath(normalized);
                if (file != null) {
                    file.refresh(false, false);
                }
            });
        } catch (Exception ignored) {
        }
    }

    private File resolve(String workspaceRoot, String path) {
        File file = new File(path);
        return file.isAbsolute() ? file : new File(workspaceRoot, path);
    }

    private boolean isBinary(File file) {
        String ext = file.getName().contains(".") ? file.getName().substring(file.getName().lastIndexOf('.') + 1).toLowerCase(Locale.ROOT) : "";
        if (BINARY_EXTS.contains(ext)) {
            return true;
        }
        byte[] bytes;
        try {
            bytes = Files.readAllBytes(file.toPath());
        } catch (Exception e) {
            return false;
        }
        int len = Math.min(bytes.length, 4096);
        if (len == 0) {
            return false;
        }
        int suspicious = 0;
        for (int i = 0; i < len; i++) {
            byte value = bytes[i];
            if (value == 0) {
                return true;
            }
            if (value < 9 || (value >= 14 && value <= 31)) {
                suspicious++;
            }
        }
        return suspicious > len * 0.3;
    }

    private record PathWithTime(long mtime, String path) {
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/tools/LiteFileTools.kt
package com.zzf.rikki.idea.agent.tools

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.vfs.LocalFileSystem
import java.io.File
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.charset.StandardCharsets
import java.util.regex.Pattern
import java.util.regex.PatternSyntaxException

/** File system tools: read, write, edit, delete, glob, grep, list. */
class LiteFileTools(private val mapper: ObjectMapper) {

    companion object {
        private const val DEFAULT_READ_LIMIT = 2_000
        private const val MAX_LINE_LEN = 2_000
        private const val MAX_BYTES = 50 * 1024
        private const val MAX_OUTPUT = 8_000
        private val BINARY_EXTS = setOf("zip", "tar", "gz", "exe", "dll", "so",
            "class", "jar", "war", "7z", "bin", "dat", "png", "jpg", "jpeg",
            "gif", "bmp", "ico", "pdf", "mp3", "mp4", "avi")
    }

    // ── read ──────────────────────────────────────────────────────────────────

    fun read(args: JsonNode, workspaceRoot: String): String {
        val pathStr = args.path("filePath").asText("")
        val offset  = args.path("offset").asInt(0).coerceAtLeast(0)
        val limit   = args.path("limit").asInt(DEFAULT_READ_LIMIT).coerceAtLeast(1)

        val file = resolve(workspaceRoot, pathStr)
        if (!file.exists()) throw RuntimeException("File not found: $pathStr")
        if (isBinary(file)) throw RuntimeException("Cannot read binary file: $pathStr")

        val allLines = file.readLines(StandardCharsets.UTF_8)
        val sb = StringBuilder("<file>\n")
        var bytes = 0
        var truncatedByBytes = false
        val end = minOf(allLines.size, offset + limit)

        for (i in offset until end) {
            val line = allLines[i].let { if (it.length > MAX_LINE_LEN) it.take(MAX_LINE_LEN) + "..." else it }
            val size = line.toByteArray(StandardCharsets.UTF_8).size + 1
            if (bytes + size > MAX_BYTES) { truncatedByBytes = true; break }
            sb.append(String.format("%05d| %s\n", i + 1, line))
            bytes += size
        }

        val lastLine = offset + (sb.lines().size - 1)  // approximate
        val totalLines = allLines.size
        when {
            truncatedByBytes -> sb.append("\n\n(Truncated at $MAX_BYTES bytes. Use 'offset' to continue.)")
            totalLines > offset + limit -> sb.append("\n\n(More lines exist. Use 'offset' to read beyond line ${offset + limit}.)")
            else -> sb.append("\n\n(End of file – $totalLines lines total)")
        }
        sb.append("\n</file>")
        return sb.toString()
    }

    // ── write ─────────────────────────────────────────────────────────────────

    fun write(args: JsonNode, workspaceRoot: String): String {
        val pathStr = args.path("filePath").asText("")
        val content = args.path("content").asText("")
        val file = resolve(workspaceRoot, pathStr)
        file.parentFile?.mkdirs()
        file.writeText(content, StandardCharsets.UTF_8)
        refreshVfsSync(file.absolutePath)
        return "Written: $pathStr (${content.lines().size} lines)"
    }

    // ── edit ──────────────────────────────────────────────────────────────────

    fun edit(args: JsonNode, workspaceRoot: String): String {
        val pathStr    = args.path("filePath").asText("")
        val oldString  = args.path("oldString").asText("")
        val newString  = args.path("newString").asText("")
        val replaceAll = args.path("replaceAll").asBoolean(false)

        val file = resolve(workspaceRoot, pathStr)

        if (!file.exists()) {
            if (oldString.isNotEmpty()) throw RuntimeException("File not found: $pathStr. To create, leave oldString empty.")
            file.parentFile?.mkdirs()
            file.writeText(newString, StandardCharsets.UTF_8)
            refreshVfsSync(file.absolutePath)
            return "Created: $pathStr"
        }

        if (oldString == newString) throw RuntimeException("oldString and newString must differ")

        val original = file.readText(StandardCharsets.UTF_8)
        val matches  = findMatches(original, oldString)
        if (matches.isEmpty()) throw RuntimeException("oldString not found in file (exact or trimmed match)")
        if (!replaceAll && matches.size > 1) throw RuntimeException(
            "oldString found ${matches.size} times — provide more context to uniquely identify the match"
        )

        val updated = if (replaceAll) matches.fold(original) { acc, m -> acc.replace(m, newString) }
                      else original.replace(matches[0], newString)

        file.writeText(updated, StandardCharsets.UTF_8)
        refreshVfsSync(file.absolutePath)
        return "Updated: $pathStr"
    }

    /** Fuzzy-match: trim each line when comparing, return the original substring. */
    private fun findMatches(content: String, find: String): List<String> {
        if (find.isEmpty()) return emptyList()
        val results = mutableListOf<String>()
        val origLines   = content.split("\n", limit = Int.MAX_VALUE)
        val searchLines = find.split("\n", limit = Int.MAX_VALUE)
            .let { lines -> if (lines.last().isEmpty()) lines.dropLast(1) else lines }
        if (searchLines.isEmpty()) return results

        for (i in 0..origLines.size - searchLines.size) {
            val matches = searchLines.indices.all { j ->
                origLines[i + j].trim() == searchLines[j].trim()
            }
            if (matches) {
                // Reconstruct the exact original substring
                var start = origLines.take(i).sumOf { it.length + 1 }
                var end   = start
                for (k in searchLines.indices) {
                    end += origLines[i + k].length
                    if (k < searchLines.size - 1) end += 1
                }
                results.add(content.substring(start, end))
            }
        }
        return results
    }

    // ── delete ────────────────────────────────────────────────────────────────

    fun delete(args: JsonNode, workspaceRoot: String): String {
        val pathStr = args.path("filePath").asText("")
        val file = resolve(workspaceRoot, pathStr)
        if (!file.exists()) throw RuntimeException("File not found: $pathStr")
        file.delete()
        refreshVfsSync(file.absolutePath)
        return "Deleted: $pathStr"
    }

    // ── glob ──────────────────────────────────────────────────────────────────

    fun glob(args: JsonNode, workspaceRoot: String): String {
        val pattern    = args.path("pattern").asText("").ifBlank { throw IllegalArgumentException("pattern required") }
        val searchRoot = resolve(workspaceRoot, args.path("path").asText(workspaceRoot))

        val matcher = FileSystems.getDefault().getPathMatcher("glob:$pattern")
        val results = mutableListOf<Pair<Long, String>>()

        Files.walk(searchRoot.toPath()).use { stream ->
            stream.filter { p ->
                val rel = searchRoot.toPath().relativize(p)
                matcher.matches(rel) || matcher.matches(p.fileName)
            }.forEach { p ->
                results += Pair(p.toFile().lastModified(), p.toString())
            }
        }

        results.sortByDescending { it.first }
        val lines = results.take(100).map { it.second }
        return if (lines.isEmpty()) "No files matched pattern: $pattern"
               else lines.joinToString("\n")
    }

    // ── grep ──────────────────────────────────────────────────────────────────

    fun grep(args: JsonNode, workspaceRoot: String): String {
        val patternStr = args.path("pattern").asText("").ifBlank { throw IllegalArgumentException("pattern required") }
        val searchRoot = resolve(workspaceRoot, args.path("path").asText(workspaceRoot))
        val include    = args.path("include").asText("").let { if (it.isBlank()) null else it }

        val regex = try { Pattern.compile(patternStr) }
                    catch (_: PatternSyntaxException) { Pattern.compile(Pattern.quote(patternStr)) }

        val includeGlob = include?.let { FileSystems.getDefault().getPathMatcher("glob:$it") }

        val results = mutableListOf<Pair<Long, String>>()

        Files.walk(searchRoot.toPath()).use { stream ->
            stream.filter { p ->
                Files.isRegularFile(p) &&
                !isBinary(p.toFile()) &&
                (includeGlob == null || includeGlob.matches(p.fileName))
            }.forEach { p ->
                try {
                    val mtime = p.toFile().lastModified()
                    p.toFile().bufferedReader(StandardCharsets.UTF_8).use { br ->
                        var lineNum = 0
                        br.forEachLine { line ->
                            lineNum++
                            if (regex.matcher(line).find()) {
                                results += Pair(mtime, "${p}:$lineNum: $line")
                            }
                        }
                    }
                } catch (_: Exception) {}
            }
        }

        results.sortByDescending { it.first }
        val lines = results.take(100).map { it.second }
        return if (lines.isEmpty()) "No matches for pattern: $patternStr"
               else lines.joinToString("\n").take(MAX_OUTPUT)
    }

    // ── list ──────────────────────────────────────────────────────────────────

    fun list(args: JsonNode, workspaceRoot: String): String {
        val dir = resolve(workspaceRoot, args.path("path").asText(workspaceRoot))
        if (!dir.isDirectory) throw RuntimeException("Not a directory: ${dir.path}")

        val ignorePatterns = buildList {
            if (args.has("ignore") && args.path("ignore").isArray) {
                args.path("ignore").forEach { add(it.asText("")) }
            }
            // Default ignores
            addAll(listOf(".git", "node_modules", ".gradle", "build", "out", ".idea", "__pycache__"))
        }

        fun shouldIgnore(name: String) = ignorePatterns.any { p -> name == p || name.startsWith(".") && p == ".*" }

        val sb = StringBuilder()
        fun walk(f: File, indent: String) {
            if (sb.length > MAX_OUTPUT) return
            val children = f.listFiles()?.sortedWith(compareBy({ !it.isDirectory }, { it.name })) ?: return
            for (child in children) {
                if (shouldIgnore(child.name)) continue
                sb.append("$indent${child.name}${if (child.isDirectory) "/" else ""}\n")
                if (child.isDirectory) walk(child, "$indent  ")
            }
        }

        walk(dir, "")
        return if (sb.isEmpty()) "Empty directory" else sb.toString().trim()
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private fun refreshVfsSync(absPath: String) {
        try {
            val normalized = absPath.replace('\\', '/')
            ApplicationManager.getApplication().invokeAndWait {
                LocalFileSystem.getInstance().refreshAndFindFileByPath(normalized)?.refresh(false, false)
            }
        } catch (_: Exception) {}
    }

    private fun resolve(workspaceRoot: String, path: String): File {
        val f = File(path)
        return if (f.isAbsolute) f else File(workspaceRoot, path)
    }

    private fun isBinary(file: File): Boolean {
        val ext = file.extension.lowercase()
        if (ext in BINARY_EXTS) return true
        val bytes = try { file.readBytes().take(4096).toByteArray() } catch (_: Exception) { return false }
        if (bytes.isEmpty()) return false
        return bytes.any { it == 0.toByte() } ||
               bytes.count { (it < 9 || it in 14..31) } > bytes.size * 0.3
    }
}
